src/net.cpp

```diff
@@ -731,6 +889,97 @@ void V1TransportSerializer::prepareForTransport(CSerializedNetMsg& msg, std::vec
+{
+    size_t serialized_command_size = 1; // short-IDs are 1 byte
+    uint8_t cmd_short_id = GetShortCommandIDFromCommand(msg.command);
+    if (cmd_short_id == 0) {
+        // message command without an assigned short-ID
+        assert(msg.command.size() <= 12);
+        // encode as varstr, max 12 chars
+        serialized_command_size = ::GetSerializeSize(msg.command, PROTOCOL_VERSION);
+    }
+    // prepare the packet length that will later be encrypted and part of the MAC (AD)
+    // the packet length excludes the 16 byte MAC tag
+    uint32_t packet_length = serialized_command_size + msg.data.size();
+
+    // prepare the packet length & message command
+    std::vector<unsigned char> serialized_header;
+    // reserve 4 bytes (3bytes AD + 1byte short-ID)
+    serialized_header.resize(CHACHA20_POLY1305_AEAD_AAD_LEN + 1);
+    // LE serialize the 24bits length
+    // we do "manually" encode this since there is no helper for 24bit serialization
+    packet_length = htole32(packet_length);
+    memcpy(serialized_header.data(), &packet_length, 3);
+
+    // append the short-ID (or eventually) the varstr of the command
+    CVectorWriter vector_writer(SER_NETWORK, INIT_PROTO_VERSION, serialized_header, 3);
+    if (cmd_short_id) {
+        // append the single byte short ID...
+        vector_writer << cmd_short_id;
+    } else {
+        // or the ASCII command string
+        vector_writer << msg.command;
+    }
+
+    // insert header directly into the CSerializedNetMsg data buffer (insert at begin)
+    // TODO: if we refactor the ChaCha20Poly1350 crypt function to allow sepearte buffers for
+    //       the AD, payload and MAC, we could avoid a insert and thus a potential reallocation
+    msg.data.insert(msg.data.begin(), serialized_header.begin(), serialized_header.end());
+
+    // resize the message buffer to make space for the MAC tag
+    msg.data.resize(msg.data.size() + 16, 0);
+
+    // length is only allowed up to 2^23 (bit24 is used for indicating rekey)
+    bool bit24 = msg.data[2] & (1u << 7);
+    assert(!bit24);
+
+    // check if we should rekey after this message
+    int64_t now = GetTime(); //TODO: check how expansive the GetTime call is and if it is avoidable
+    bool rekey = false;
+    if (m_bytes_encrypted >= (gArgs.GetBoolArg("-netencryptionfastrekey", false) ? 32 * 1024 : REKEY_LIMIT_BYTES) || now - m_time_last_rekey >= (gArgs.GetBoolArg("-netencryptionfastrekey", false) ? 10 : REKEY_LIMIT_TIME)) {
+        LogPrint(BCLog::NET, "Rekey limits reached, performing rekey.\n");
+        msg.data[2] |= (1u << 7);
+        rekey = true;
+    }
+
+    // encrypt the payload, ignore return code since it can't fail in this case (controlled buffers, don't check the MAC during encrypting)
+    m_aead->Crypt(m_payload_seqnr, m_aad_seqnr, m_aad_pos, msg.data.data(), msg.data.size(), msg.data.data(), msg.data.size() - 16, true);
```

Comment:
Mind using a named constant:
```suggestion
    m_aead->Crypt(m_payload_seqnr, m_aad_seqnr, m_aad_pos, msg.data.data(), msg.data.size(), msg.data.data(), msg.data.size() - CHACHA20_POLY1305_AEAD_TAG_LEN, true);
```
?